# Predicting Industrial Machine Downtime: Level 1

## 📖 Background
You work for a manufacturer of high-precision metal components used in aerospace, automotives, and medical device applications. Your company operates three different machines on its shop floor that produce different sized components, so minimizing the downtime of these machines is vital for meeting production deadlines.

Your team wants to use a data-driven approach to predicting machine downtime, so proactive maintenance can be planned rather than being reactive to machine failure. To support this, your company has been collecting operational data for over a year and whether each machine was down at those times.

In this first level, you're going to explore and describe the data. This level is aimed towards beginners. If you want to challenge yourself a bit more, check out level two!

## 💾 The data

The company has stored the machine operating data in a single table, available in `'data/machine_downtime.csv'`.

#### Each row in the table represents the operational data for a single machine on a given day:
- `"Date"` - the date the reading was taken on.
- `"Machine_ID"` - the unique identifier of the machine being read.
- `"Assembly_Line_No"` - the unique identifier of the assembly line the machine is located on.
- `"Hydraulic_Pressure(bar)"`, `"Coolant_Pressure(bar)"`, and `"Air_System_Pressure(bar)"` - pressure measurements at different points in the machine.
- `"Coolant_Temperature"`, `"Hydraulic_Oil_Temperature"`, and `"Spindle_Bearing_Temperature"` - temperature measurements (in Celsius) at different points in the machine.
- `"Spindle_Vibration"`, `"Tool_Vibration"`, and `"Spindle_Speed(RPM)"` - vibration (measured in micrometers) and rotational speed measurements for the spindle and tool.
- `"Voltage(volts)"` - the voltage supplied to the machine.
- `"Torque(Nm)"` - the torque being generated by the machine.
- `"Cutting(KN)"` - the cutting force of the tool.
- `"Downtime"` - an indicator of whether the machine was down or not on the given day.

In [1]:
# import the necessary libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')

## Data Analysis

In [2]:
# read the data and print the first few rows
machine_downtime = pd.read_csv('C://Users/Administrator/Documents/Data Science Projects/Machine-Downtime-Prediction/data/machine_downtime.csv')
machine_downtime.head()

Date            Machine_ID Assembly_Line_No  Hydraulic_Pressure(bar)  \
0  31-12-2021  Makino-L1-Unit1-2013     Shopfloor-L1                    71.04   
1  31-12-2021  Makino-L1-Unit1-2013     Shopfloor-L1                   125.33   
2  31-12-2021  Makino-L3-Unit1-2015     Shopfloor-L3                    71.12   
3  31-05-2022  Makino-L2-Unit1-2015     Shopfloor-L2                   139.34   
4  31-03-2022  Makino-L1-Unit1-2013     Shopfloor-L1                    60.51   

   Coolant_Pressure(bar)  Air_System_Pressure(bar)  Coolant_Temperature  \
0               6.933725                  6.284965                 25.6   
1               4.936892                  6.196733                 35.3   
2               6.839413                  6.655448                 13.1   
3               4.574382                  6.560394                 24.4   
4               6.893182                  6.141238                  4.1   

   Hydraulic_Oil_Temperature  Spindle_Bearing_Temperature  Spindle_Vibration  \
0                       46.0                         33.4              1.291   
1                       47.4                         34.6              1.382   
2                       40.7                         33.0              1.319   
3                       44.2                         40.6              0.618   
4                       47.3                         31.4              0.983   

   Tool_Vibration  Spindle_Speed(RPM)  Voltage(volts)  Torque(Nm)  \
0          26.492             25892.0           335.0   24.055326   
1          25.274             19856.0           368.0   14.202890   
2          30.608             19851.0           325.0   24.049267   
3          30.791             18461.0           360.0   25.860029   
4          25.516             26526.0           354.0   25.515874   

   Cutting(kN)         Downtime  
0         3.58  Machine_Failure  
1         2.68  Machine_Failure  
2         3.55  Machine_Failure  
3         3.55  Machine_Failure  
4         3.55  Machine_Failure

## Data Wrangling

In [3]:
# create a copy of the original data
machine_downtime_ori = machine_downtime.copy()

### 1. inconsistent data types and NaN values 
+ The `Date` column is type **object** instead of type **datetime**
+ There are NaN values in majority of the columns. we need to do further analysis on that
+ The other column data types are okay


In [12]:
# get data types of columns
machine_downtime.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2500 entries, 0 to 2499
Data columns (total 16 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Date                         2500 non-null   object 
 1   Machine_ID                   2500 non-null   object 
 2   Assembly_Line_No             2500 non-null   object 
 3   Hydraulic_Pressure(bar)      2490 non-null   float64
 4   Coolant_Pressure(bar)        2481 non-null   float64
 5   Air_System_Pressure(bar)     2483 non-null   float64
 6   Coolant_Temperature          2488 non-null   float64
 7   Hydraulic_Oil_Temperature    2484 non-null   float64
 8   Spindle_Bearing_Temperature  2493 non-null   float64
 9   Spindle_Vibration            2489 non-null   float64
 10  Tool_Vibration               2489 non-null   float64
 11  Spindle_Speed(RPM)           2494 non-null   float64
 12  Voltage(volts)               2494 non-null   float64
 13  Torque(Nm)        

### 2. Value range constraint 
+ There are negative values in `Hydraulic_pressure` and `Spindle_Vibration` column. we will investigate that
+ We would convert the `cutting(kN)` SI unit column to **newton(N)** from **KiloNewton(kN)** and the `Spindle_Speed` from **Rotation per minute (RPM)** to **Rotation per seconds(RPS)** since all other variable are in their standard units
+ for the `Date` column, we will drill into it more during the data correction phase




In [15]:
# get the summray statistics of the data
machine_downtime_ori.describe()

Hydraulic_Pressure(bar)  Coolant_Pressure(bar)  \
count              2490.000000            2481.000000   
mean                101.409084               4.947058   
std                  30.289301               0.997357   
min                 -14.326454               0.325000   
25%                  76.355000               4.462464   
50%                  96.763486               4.939960   
75%                 126.415727               5.518401   
max                 191.000000              11.350000   

       Air_System_Pressure(bar)  Coolant_Temperature  \
count               2483.000000          2488.000000   
mean                   6.499275            18.559887   
std                    0.407279             8.554480   
min                    5.063480             4.100000   
25%                    6.217987            10.400000   
50%                    6.505191            21.200000   
75%                    6.780550            25.600000   
max                    7.973992            98.200000   

       Hydraulic_Oil_Temperature  Spindle_Bearing_Temperature  \
count                2484.000000                  2493.000000   
mean                   47.618317                    35.063698   
std                     3.768674                     3.764823   
min                    35.200000                    22.600000   
25%                    45.100000                    32.500000   
50%                    47.700000                    35.100000   
75%                    50.100000                    37.600000   
max                    61.400000                    49.500000   

       Spindle_Vibration  Tool_Vibration  Spindle_Speed(RPM)  Voltage(volts)  \
count        2489.000000     2489.000000         2494.000000     2494.000000   
mean            1.009334       25.411975        20274.792302      348.996792   
std             0.342898        6.437142         3852.660056       45.376024   
min            -0.461000        2.161000            0.000000      202.000000   
25%             0.777000       21.089000        17919.000000      319.000000   
50%             1.008000       25.455000        20137.500000      349.000000   
75%             1.236000       29.790000        22501.750000      380.000000   
max             2.000000       45.726000        27957.000000      479.000000   

        Torque(Nm)  Cutting(kN)  
count  2479.000000  2493.000000  
mean     25.234968     2.782551  
std       6.138564     0.616688  
min       0.000000     1.800000  
25%      21.666115     2.250000  
50%      24.647736     2.780000  
75%      30.514008     3.270000  
max      55.552400     3.930000

### 3. Uniqueness Constraints
+ No spelling irregularities in the `Downtime` column
+ There are no duplicates present


In [7]:
# check the values of `downtime` column
machine_downtime_ori.Downtime.unique()

array(['Machine_Failure', 'No_Machine_Failure'], dtype=object)

In [10]:
# check for duplicates
machine_downtime_ori.duplicated().sum()

np.int64(0)

## 💪 Competition challenge

Create a report that covers the following:
1. What is the first and last date readings were taken on?
2. What is the average Torque?
3. Which assembly line has the highest readings of machine downtime?

## 🧑‍⚖️ Judging criteria

| CATEGORY | WEIGHTING | DETAILS                                                              |
|:---------|:----------|:---------------------------------------------------------------------|
| **Recommendations** | 35%       | <ul><li>Clarity of recommendations - how clear and well presented the recommendation is.</li><li>Quality of recommendations - are appropriate analytical techniques used & are the conclusions valid?</li><li>Number of relevant insights found for the target audience.</li></ul>       |
| **Storytelling**  | 35%       | <ul><li>How well the data and insights are connected to the recommendation.</li><li>How the narrative and whole report connects together.</li><li>Balancing making the report in-depth enough but also concise.</li></ul> |
| **Visualizations** | 20% | <ul><li>Appropriateness of visualization used.</li><li>Clarity of insight from visualization.</li></ul> |
| **Votes** | 10% | <ul><li>Up voting - most upvoted entries get the most points.</li></ul> |

## ✅ Checklist before publishing into the competition
- Rename your workspace to make it descriptive of your work. N.B. you should leave the notebook name as notebook.ipynb.
- **Remove redundant cells** like the judging criteria, so the workbook is focused on your story.
- Make sure the workbook reads well and explains how you found your insights. 
- Try to include an **executive summary** of your recommendations at the beginning.
- Check that all the cells run without error

## ⌛️ Time is ticking. Good luck!